ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/Operation.java

```diff
@@ -20,34 +20,33 @@
 
 public interface Operation {
 
-  /**
-   * Gas cost of this operation, in context of the provided frame.
-   *
-   * @param frame The frame for execution of this operation.
-   * @return The gas cost associated with executing this operation given the current {@link
-   *     MessageFrame}.
-   */
-  Gas cost(MessageFrame frame);
+  class OperationResult {
+    final Optional<Gas> gasCost;
+    final Optional<ExceptionalHaltReason> haltReason;
+
+    public OperationResult(
+        final Optional<Gas> gasCost, final Optional<ExceptionalHaltReason> haltReason) {
+      this.gasCost = gasCost;
+      this.haltReason = haltReason;
+    }
+
+    public Optional<Gas> getGasCost() {
+      return gasCost;
+    }
+
+    public Optional<ExceptionalHaltReason> getHaltReason() {
+      return haltReason;
+    }
+  }
 
   /**
    * Executes the logic behind this operation.
    *
    * @param frame The frame for execution of this operation.
+   * @param evm The EVM for execution of this operation.
+   * @return the gas cost and any exeptional halt reasons of the operation.
    */
-  void execute(MessageFrame frame);
-
-  /**
-   * Check if an exceptional halt condition should apply
-   *
-   * @param frame the current frame
-   * @param evm the currently executing EVM
-   * @return an {@link Optional} containing the {@link ExceptionalHaltReason} that applies or empty
-   *     if no exceptional halt condition applies.
-   */
-  default Optional<ExceptionalHaltReason> exceptionalHaltCondition(
-      final MessageFrame frame, final EVM evm) {
-    return Optional.empty();
-  }
+  OperationResult execute(final MessageFrame frame, final EVM evm);
```

Comment:
Since this now also checks gas, we should reflect the implementer's responsibility in the javadoc if not the name of the method.

Reply:
javadocs updated.